import subprocess
import os
import sys
import glob
import yaml 

##### load paths #####
path_folder_prefix="/tarafs/scratch/proj5034-AGBKU/catfish"
###
path_reads_prefix=f"{path_folder_prefix}/00-DATA/00-READS"
path_data_prefix=f"{path_folder_prefix}/00-DATA/"
path_prog_prefix=f"{path_folder_prefix}/01-PROGRAMS"
path_exp_prefix=f"{path_folder_prefix}/02-EXPERIMENTS"
path_out_prefix=f"{path_folder_prefix}/03-OUTPUTS"
path_tmp_prefix=f"{path_folder_prefix}/04-TEMPORARY"  

##### setup config and report #####
report: "report/workflow.rst"
threads = 12
configfile: "config/config.yaml"

##### Definition of functions

def get_unique_values(key):
    values = set()
    for sample in config["samples"].values():
        values.update(sample.get(key, []))
    return list(values)

def get_sample_ids():
    return get_unique_values('sample_ids')

def get_reference_accession():
    return get_unique_values('refseq_genbank_accession')

def get_species_file_prefix():
    return get_unique_values('file_prefixes')

def get_species_long():
    return get_unique_values('species_long')

def get_species():
    return get_unique_values('species')

def get_sex():
    return get_unique_values('sex')

def get_method():
    return get_unique_values('method')

def get_orientation():
    return get_unique_values('orientation')

def get_combined_samples():
    species = get_species()
    sex = get_sex()
    method = get_method()
    orientation = get_orientation()
    return ["{}_{}_{}_{}".format(s, sx, m, o) for s in species for sx in sex for m in method for o in orientation]

def get_sample(wildcards): #Get_sample_prefix
    # This function now assumes each sample is a key in the config that 
    # corresponds to a prefix without a file extension
    return f"{wildcards.species}_{wildcards.sex}_{wildcards.method}_{wildcards.orientation}"

##### load init rules #####

rule all:
    input:
        "logs/test.txt",
        expand(f"{path_out_prefix}/00-MASH/{{sample}}_combined.tbl", path_out_prefix=path_out_prefix, sample=get_combined_samples()),
        expand(f"{path_out_prefix}/00-MASH/{{sample}}_screen.out", path_out_prefix=path_out_prefix, sample=get_combined_samples()),
        expand(f"{path_out_prefix}/00-MASH/{{sample}}_mash_plot.png", path_out_prefix=path_out_prefix, sample=get_combined_samples())
        # Uncomment the following lines when you want to include them in the execution
        # expand("{path_out_prefix}/00-JELLYFISH/{sample}_jellyfishcount.jf", sample=get_combined_samples()),
        # expand("{path_out_prefix}/00-JELLYFISH/{sample}_jellyfish_histo.histo", sample=get_combined_samples()),
        # expand("{path_out_prefix}/00-GENOMESCOPE/{sample}_plot.png", sample=get_combined_samples())
        # Add additional lines for other output files as needed

##### Main rule to orchestrate the workflow #####

##### load rules #####
#include: "rules/plots.smk"
#include: "rules/reference.smk"
include: "rules/quality_control_reads.smk"
#include: "rules/quality_control_reads_jellyfish.smk"
#include: "rules/quality_control_assembly.smk" 
#include: "rules/trimming.smk"
#include: "rules/denovo.smk"

rule create_log_folder:
    output:
	"logs/test.txt"
    shell: "touch {output}"


#rule create_assembly_metadata:
#    input:
#        expand("path/to/assemblies/{sample}_assembly.fasta", sample=SAMPLES)
#    output:
#        "assemblies_metadata.yaml"
#    run:
#        import yaml
#
#        assembly_data = {}
#        for sample in SAMPLES:
#            assembly_data[sample] = {"assembly": f"path/to/assemblies/{sample}_assembly.fasta"}
#        
#        with open(output[0], 'w') as outfile:
#            yaml.dump(assembly_data, outfile, default_flow_style=False)



#include: "rules/mapping_illumina.smk"
#include: "rules/variant_calling.smk"

#include: "rules/pangenomes_plots.smk"

#include: "rules/mitoVGP_conda_env_ONT.smk"
#include: "rules/mitoVG_conda_env_pacbio.smk"


#include: "rules/annotations.smk"
#include: "rules/pangenomes_annotations.smk"

#rule create_annotations_metadata:
#    input:
#        expand("path/to/assemblies/{sample}_assembly.fasta", sample=SAMPLES)
#    output:
#        "annotations_metadata.yaml"
#    run:
#        import yaml
#
#        annotation_data = {}
#        for sample in SAMPLES:
#            annotation_data[sample] = {"assembly": f"path/to/assemblies/{sample}_assembly.fasta"}
#        
#        with open(output[0], 'w') as outfile:
#            yaml.dump(annotation_data, outfile, default_flow_style=False)
#
#
# End
