The provided workflow is a set of bioinformatic rules structured for Snakemake, a workflow management system that orchestrates data processing in a scalable and reproducible manner. It begins with the preprocessing of different types of sequencing data: Nanopore, Illumina, Hi-C, and PacBio. Each data type undergoes quality control, with FastQC used for assessing raw data quality, followed by adapter trimming where necessary. For HiFi reads from the PacBio sequencing, additional steps for format conversion and adapter removal are executed. After preprocessing, the workflow includes a 'mash' step that sketches the reads to create compressed representations, which are then used in 'mash screen' to screen against a reference database. This is followed by 'mash dist' to calculate pairwise distances between the sketches and reference sequences, aiding in the identification of closely related sequences. The distances are further visualized in a plot generated by an R script. Logging is integrated at various steps to facilitate tracking and debugging of the workflow. This modular approach allows for efficient and comprehensive analysis, from raw reads to data interpretation, tailored for genome assembly and pathogen identification in aquaculture species.
